aptos-move/e2e-move-tests/src/tests/aggregator_v2.rs

```diff
@@ -100,105 +104,214 @@ fn test_aggregators_e2e() {
+) {
+    assert!(first + second <= txn_block.len());
+    let (left, rest) = txn_block.split_at(first);
+    let (mid, right) = rest.split_at(second);
+
+    fn run_and_check_block(
+        harness: &mut MoveHarness,
+        txn_block: Vec<(u64, SignedTransaction)>,
+        offset: usize,
+    ) {
+        if txn_block.is_empty() {
+            return;
+        }
+        let (errors, txns): (Vec<_>, Vec<_>) = txn_block.into_iter().unzip();
+        let outputs = harness.run_block(txns);
+        for (idx, (error, status)) in errors.into_iter().zip(outputs.into_iter()).enumerate() {
+            if error > 0 {
+                // TODO update to assert_error and assert_success after changes there land
+                assert!(
+                    matches!(
+                        status,
+                        aptos_types::transaction::TransactionStatus::Keep(
+                            aptos_types::transaction::ExecutionStatus::MoveAbort { code, .. }
+                        ) if code == error
+                    ),
+                    "Failed on txn {}, with block starting at {}",
+                    idx + offset,
+                    offset
+                );
+            } else {
+                assert_eq!(
+                    status,
+                    aptos_types::transaction::TransactionStatus::Keep(
+                        aptos_types::transaction::ExecutionStatus::Success
+                    ),
+                    "Succeeded on txn {} that should've failed, with block starting at {}",
+                    idx + offset,
+                    offset
+                );
+            }
+        }
+    }
+
+    run_and_check_block(harness, left.to_vec(), 0);
+    run_and_check_block(harness, mid.to_vec(), first);
+    run_and_check_block(harness, right.to_vec(), first + second);
 }
 
-#[test]
-#[should_panic]
-fn test_aggregator_underflow() {
-    let (mut h, acc) = setup();
+fn arb_block_split(len: usize) -> BoxedStrategy<(usize, usize)> {
+    any::<bool>()
+        .prop_flat_map(move |whole| {
+            // making running a test with a full block likely
+            if whole {
+                Just((len, 0)).boxed()
+            } else {
+                // First is non-empty, and not the whole block here: [1, len)
+                (1usize..len)
+                    .prop_flat_map(move |first| {
+                        // Second is non-empty, but can finish the block: [1, len - first]
+                        (Just(first), 1usize..len - first + 1)
+                    })
+                    .boxed()
+            }
+        })
+        .boxed()
+}
 
-    let txn1 = new(&mut h, &acc, 0, 600);
-    let txn2 = try_add(&mut h, &acc, 0, 400);
-    let txn3 = try_sub(&mut h, &acc, 0, 500);
+proptest! {
+    #![proptest_config(ProptestConfig {
+        // Cases are expensive, few cases is enough.
+        cases: 5,
+        result_cache: prop::test_runner::basic_result_cache,
+        .. ProptestConfig::default()
+    })]
```

Comment:
What does this statement do?
What's the difference between the proptests and the regular Rust tests we had earlier?

Reply:
proptest allows you to run a parametrized test. 
I.e. to have a test with arguments, and proptest will run it multiple times with different arguments.
Here, I paramtrized how transactions are split into blocks - i.e. whether a whole test is run as a single block, or it is split in two or three blocks, and random positions (first_block and second_block are lenghts of each block)

This config does two things:
- makes every test run 5 times. as tests are expensive, default of 1000 for proptest is not practical here, and 5 gives enough variety
- "caches" the run, so it doesn't call the same test with same arguments twice (default is to allow calling same test with same arguments twice, for some reason)